### PR TITLE
[Feature]: Add support for DevLogin

### LINKED ISSUE #100: [Feature Request] Add DevLogin support to client run configs
[DevLogin](https://github.com/covers1624/DevLogin#devlogin) allows users to authenticate their clients using Mojangs services at startup.

Adding DevLogin to any NeoForge mod involves the [following steps currently](https://github.com/covers1624/DevLogin/wiki/NeoGradle).

While this is a simple setup, I think it would make sense for NeoGradle, and potentially other loaders too, to include DevLogin as a feature for their client run configs.
Of course this would have to be a opt-in feature, having it enabled by default wouldn't make much sense IMO.

I am thinking of adding a super simple boolean flag (`authed true`) to client configs, which when enabled
- sets the main class to `net.covers1624.devlogin.DevLogin`
- sets the program arg `--launch_target` to `cpw.mods.bootstraplauncher.BootstrapLauncher`
- and tells DevLogin to use the `default` auth profile via `--launch_profile default`
	There should also be an optional property users can specify to set the profile that DevLogin will use. This will allow users to authenticate their clients with multiple accounts.
	Maybe something like `authProfile <profile>` (if set also sets `authed true`)

### PR DESCRIPTION
**Pull request by marchermans**
### TLDR: 
- This adds supports for DevLogin on client runs

---

### DevLogin Support:
We now support the use of DevLogin by @covers1624.
You can use the following DSL to enable it:
```groovy
runs {
    someRun {
        devLogin {
           enabled true
        }
    }
}
```
By default the tool is disabled on all client runs.

> [!WARNING]
> If you enable the dev login tool for a none client run, you will get an error message.

### Better error handling:
There is now better error handling if you do not configure mod sources in a run.

### Closes:
Closes: #100

### CODE REVIEW

**Review by Technici4n — DISMISSED** [review]
Please make it disabled by default. At least in neodev.

**Comment by marchermans** [comment]
> Please make it disabled by default. At least in neodev.

Is there a reason why?
Not against this, just want to know why, so I can leave a comment explaining the need to disable it in dynprojects.

**Comment by marchermans** [comment]
> > Please make it disabled by default. At least in neodev.
> 
> Is there a reason why? Not against this, just want to know why, so I can leave a comment explaining the need to disable it in dynprojects.

Clarified via discord. Was personal preference of @Technici4n.

We decided to disable the flag on the run, this leaves the subsystem enabled, while having no effect for the user.

File: README.md [hunk 1]
```diff
@@ -359,3 +359,71 @@ The following properties can be used to configure the JST tool:
 ```properties
 neogradle.subsystems.tools.jst=<artifact coordinate for jst cli tool>
 ```
+### DevLogin
+This tool is used by the runs subsystem to enable logged in plays on all client runs.
+The following properties can be used to configure the DevLogin tool:
+```properties
+neogradle.subsystems.tools.devLogin=<artifact coordinate for devlogin cli tool>
```
Comment (sciwhiz12) [inline README.md:365]: Perhaps this should document the GitHub repository of the DevLogin tool.

File: common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java [hunk 2]
```diff
@@ -374,6 +378,30 @@ private void applyAfterEvaluate(final Project project) {
...
+                            defaultRuntimeOnlyConfiguration.getDependencies().add(project.getDependencies().create(tools.getDevLogin().get()));
+
+                            //Update the program arguments to properly launch the dev login tool
+                            run.getProgramArguments().add("--launch_target");
+                            run.getProgramArguments().add(mainClass);
+
+                            //Set the main class to the dev login tool
+                            run.getMainClass().set("net.covers1624.devlogin.DevLogin");
+                        } else if (!runImpl.getIsClient().get() && runImpl.getShouldUseDevLogin().get()) {
+                            throw new GradleException("Dev login is only supported on runs which are marked as clients! The run: " + runImpl.getName() + " is not a client run.");
```
Comment (sciwhiz12) [inline common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java]: Perhaps [`InvalidUserDataException`](https://docs.gradle.org/8.7/javadoc/org/gradle/api/InvalidUserDataException.html) would be a more fitting exception here, than the generic `GradleException`.

File: common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java [hunk 3]
```diff
@@ -374,6 +378,30 @@ private void applyAfterEvaluate(final Project project) {
...
+                            final SourceSet defaultSourceSet = runImpl.getModSources().get().get(0);
+                            final Configuration defaultRuntimeOnlyConfiguration = project.getConfigurations().maybeCreate(defaultSourceSet.getRuntimeOnlyConfigurationName());
+                            defaultRuntimeOnlyConfiguration.getDependencies().add(project.getDependencies().create(tools.getDevLogin().get()));
+
+                            //Update the program arguments to properly launch the dev login tool
+                            run.getProgramArguments().add("--launch_target");
+                            run.getProgramArguments().add(mainClass);
+
+                            //Set the main class to the dev login tool
+                            run.getMainClass().set("net.covers1624.devlogin.DevLogin");
```
Comment (sciwhiz12) [inline common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java]: Is this configurable by the end-user? As in, if they switch out the DevLogin artifact in the future, can they change the main class of the run manually, or is this fixed in NeoGradle now?

File: README.md [hunk 4]
```diff
@@ -359,3 +359,72 @@ The following properties can be used to configure the JST tool:
...
+``` 
+By default, the repositories are added, but if you want to disable this, you can set the property to false.
+By doing so you are responsible for adding a maven repository that provides the dev login CLI artifact.
+
+### Per run configuration
+If you want to configure the dev login tool per run, you can do so by setting the following properties in your run configuration:
+```groovy
+runs {
+    someRun {
+        shouldUseDevLogin true
```
Comment (sciwhiz12) [inline README.md]: Why is it called `shouldUseDevLogin`? Why not `useDevLogin`?

File: dsl/common/src/main/groovy/net/neoforged/gradle/dsl/common/util/Constants.groovy [hunk 5]
```diff
@@ -28,6 +28,11 @@ class Constants {
     public static final String DEFAULT_PARCHMENT_ARTIFACT_PREFIX = "parchment-"
     public static final String DEFAULT_PARCHMENT_MAVEN_URL = "https://maven.parchmentmc.org/"
     public static final String JST_TOOL_ARTIFACT = "net.neoforged.jst:jst-cli-bundle:1.0.36"
+    public static final String DEFAULT_DEVLOGIN_GROUP = "net.covers1624"
+    public static final String DEFAULT_DEVLOGIN_ARTIFACT = "DevLogin"
+    public static final String DEFAULT_DEVLOGIN_MAVEN_URL = "https://maven.covers1624.net/"
```
Comment (Technici4n) [inline dsl/common/src/main/groovy/net/neoforged/gradle/dsl/common/util/Constants.groovy]: Isn't it already available via our maven?

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java (hunk 3 above)
Comment (marchermans) [reply common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java]: I had it hardcoded, but your comment made me add a subsystem property to configure it,

File: README.md [hunk 6]
```diff
@@ -359,3 +359,72 @@ The following properties can be used to configure the JST tool:
...
+    }
+}
+```
+This will enable the dev login tool for this run. 
+By default, the dev login tool is disabled, and can only be enabled for client runs.
+
+> [!WARNING]
+> If you enable the dev login tool for a none client run, you will get an error message.
```
Comment (Matyrobbrt) [inline README.md]: You need the `non` prefix, not the `none` pronoun.

**Comment by marchermans** [comment]
Addressed all comments.

File: README.md [hunk 7]
```diff
@@ -359,3 +358,73 @@ The following properties can be used to configure the JST tool:
...
+``` 
+By default, the repositories are added, but if you want to disable this, you can set the property to false.
+By doing so you are responsible for adding a maven repository that provides the dev login CLI artifact.
+
+### Per run configuration
+If you want to configure the dev login tool per run, you can do so by setting the following properties in your run configuration:
+```groovy
+runs {
+    someRun {
+        shouldUseDevLogin true
```
Comment (sciwhiz12) [inline README.md]: You forgot to update this line to match the new property `useDevLogin`.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -359,3 +358,73 @@ The following properties can be used to configure the JST tool:
...
+```properties
+net.neoforged.gradle.caching.enabled=false
+```
+
+You can clean the artifacts that are stored in the cache by running the following command:
+```shell
+./gradlew cleanCache
+```
+
+This command is also automatically ran, when you run the clean task.
```
Comment (Matyrobbrt) [inline README.md]: This sentence is in the passive voice, which requires a past participle verb (`run` being the past participle of `run`)

File: README.md [hunk 9]
```diff
@@ -359,3 +358,73 @@ The following properties can be used to configure the JST tool:
...
+The command will check if the stored artifact count is higher than the configured threshold, and if so remove the oldest artifacts until the count is below the threshold.
```
Comment (Matyrobbrt) [inline README.md]:
```suggestion
The command will check if the stored artifact count is higher than the configured threshold, and if so, remove the oldest artifacts until the count is below the threshold.
```

File: README.md [hunk 10]
```diff
@@ -359,3 +358,73 @@ The following properties can be used to configure the JST tool:
...
+More information on the relevant tool, its released version and documentation can be found here: [DevLogin by Covers1624](https://github.com/covers1624/DevLogin)
+
+## DevLogin
+The DevLogin tool is a tool that allows you to log in to a Minecraft account without having to use the Minecraft launcher, during development.
+This tool is used by the runs subsystem to enable logged in plays on all client runs.
+The tool can be configured using the following properties:
+```properties
+neogradle.subsystems.devLogin.enabled=<true/false>
+```
+By default, the subsystem is enabled, and it will ask you to log in when you run a client run.
```
Comment (Matyrobbrt) [inline README.md]: This is contradicting with the lines below. Is it, or is it not enabled?

File: README.md [hunk 11]
```diff
@@ -359,3 +358,73 @@ The following properties can be used to configure the JST tool:
 ```properties
 neogradle.subsystems.tools.jst=<artifact coordinate for jst cli tool>
 ```
+### DevLogin
+This tool is used by the runs subsystem to enable logged in plays on all client runs.
```
Comment (Matyrobbrt) [inline README.md]:
```suggestion
This tool is used by the runs subsystem to enable Minecraft authentification in client runs.
```

File: README.md [hunk 12]
```diff
@@ -359,3 +358,73 @@ The following properties can be used to configure the JST tool:
...
+
+## DevLogin
+The DevLogin tool is a tool that allows you to log in to a Minecraft account without having to use the Minecraft launcher, during development.
+This tool is used by the runs subsystem to enable logged in plays on all client runs.
+The tool can be configured using the following properties:
+```properties
+neogradle.subsystems.devLogin.enabled=<true/false>
+```
+By default, the subsystem is enabled, and it will ask you to log in when you run a client run.
+If you want to disable this, you can set the property to false, and then you will not be asked to log in, but use a random none signed in dev account.
```
Comment (Matyrobbrt) [inline README.md]: No comma after if clauses. Also, `non`, not `none`.

File: common/src/main/java/net/neoforged/gradle/common/extensions/subsystems/SubsystemsExtension.java [hunk 13]
```diff
@@ -33,13 +29,27 @@ public SubsystemsExtension(Project project) {
         configureRecompilerDefaults();
         configureParchmentDefaults();
         configureToolsDefaults();
+        configureDevLoginDefaults();
+    }
+
+    private void configureDevLoginDefaults() {
+        DevLogin devLogin = getDevLogin();
+        devLogin.getEnabled().convention(
+                getBooleanProperty("devLogin.enabled").orElse(true)
```
Comment (Matyrobbrt) [inline common/src/main/java/net/neoforged/gradle/common/extensions/subsystems/SubsystemsExtension.java:38]: Wasn't it meant to be false by default?

File: common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java [hunk 14]
```diff
@@ -374,6 +378,30 @@ private void applyAfterEvaluate(final Project project) {
...
+                    final DevLogin devLogin = project.getExtensions().getByType(Subsystems.class).getDevLogin();
+                    final Tools tools = project.getExtensions().getByType(Subsystems.class).getTools();
+                    if (devLogin.getEnabled().get()) {
+                        //Dev login is only supported on the client side
+                        if (runImpl.getIsClient().get() && runImpl.getUseDevLogin().get()) {
+                            final String mainClass = runImpl.getMainClass().get();
+
+                            //We add the dev login tool to the runtime only configuration, of the first source set, this should suffice
+                            final SourceSet defaultSourceSet = runImpl.getModSources().get().get(0);
+                            final Configuration defaultRuntimeOnlyConfiguration = project.getConfigurations().maybeCreate(defaultSourceSet.getRuntimeOnlyConfigurationName());
```
Comment (Matyrobbrt) [inline common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java]: `runtimeOnly` is a leaking configuration, please user another one that inherits from `runtimeClasspath`.

**Review by Matyrobbrt — DISMISSED** [review]

File: common/src/main/java/net/neoforged/gradle/common/extensions/subsystems/SubsystemsExtension.java (hunk 13 above)
Comment (marchermans) [reply common/src/main/java/net/neoforged/gradle/common/extensions/subsystems/SubsystemsExtension.java:38]: By lukes suggestion, and the one I agreed with, the subsystem it self is enabled, yet the default on each run is `false` so the user only needs to enable the relevant runs he or she wants.

File: common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java (hunk 14 above)
Comment (marchermans) [reply common/src/main/java/net/neoforged/gradle/common/CommonProjectPlugin.java]: Addressed, made a custom runtime only configuration and extended it in the same way that the localRuntime does when that is enabled.

File: README.md [hunk 15]
```diff
@@ -359,3 +358,76 @@ The following properties can be used to configure the JST tool:
...
+### DevLogin
+This tool is used by the runs subsystem to enable Minecraft authentication in client runs.
+The following properties can be used to configure the DevLogin tool:
+```properties
+neogradle.subsystems.tools.devLogin=<artifact coordinate for devlogin cli tool>
+```
+More information on the relevant tool, its released version and documentation can be found here: [DevLogin by Covers1624](https://github.com/covers1624/DevLogin)
+
+## DevLogin
+The DevLogin tool is a tool that allows you to log in to a Minecraft account without having to use the Minecraft launcher, during development.
```
Comment (sciwhiz12) [inline README.md:370]: As I understand it, DevLogin works when the developer is logged into the vanilla Minecraft launcher, correct? That should be clarified here.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: README.md [hunk 16]
```diff
@@ -359,3 +358,83 @@ The following properties can be used to configure the JST tool:
...
+By default, it will use the credentials stored by the vanilla launcher, however if you are not logged in there, it will offer you the ability to log in directly from the console.
```
Comment (covers1624) [inline README.md]: DevLogin does not have the ability to use the vanilla launcher credentials. It purely implements the device auth flow. If someone can figure out how they store the creds, we can absolutely add support for it.

**Comment by covers1624** [comment]
Glad to see a tool I wrote to give me my skin in-dev put to real use.

It may be worth exposing DevLogin's profile system via the DSL `--launch_profile`, so modders can easily use their alts. Just a simple identifier string for the profile.

**Review by sciwhiz12 — APPROVED** [review]
